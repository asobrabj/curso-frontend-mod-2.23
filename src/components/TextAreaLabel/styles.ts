import styled from 'styled-components'

interface ITextAreaLabelStyle {
  $fs?: 'xsmall' | 'small' | 'medium' | 'larger' | string
  $fStyle?: 'normal' | 'italic'
  $fw?: 'medium' | 'larger'
  $fFamily?: string
  $tAlign?: 'start' | 'end' | 'justify' | 'center'
  $tTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase'
  $color?: string
  $bgColor?: string
  $pd?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  $pdTop?: string
  $pdBottom?: string
  $pdRigth?: string
  $pdLeft?: string
  $mg?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  $mgTop?: string
  $mgBottom?: string
  $mgRigth?: string
  $mgLeft?: string
  $width?: string
  $height?: string
  $resize?: 'both' | 'none' | 'vertical' | 'horizontal'
  $bdColor?: string
}

interface Defaults {
  [key: string]: string
}

const fontSize = (fs: string): string => {
  const defaults: Defaults = {
    xsmall: '14',
    small: '16',
    medium: '24',
    larger: '32',
  }
  const value = defaults[`${fs}`] || fs
  return `${value}px`
}

const fontWheigt = (fw: string): string => {
  const defaults: Defaults = {
    medium: '400',
    larger: 'bold',
  }
  const value = defaults[`${fw}`]
  return value
}

const paddingFn = (pd: string): string => {
  const defaults: Defaults = {
    xsmall: '5',
    small: '8',
    medium: '16',
    larger: '24',
    xlarger: '32',
  }
  const value = defaults[`${pd}`] ? `${defaults[`${pd}`]}px` : ''
  return value
}

const marginFn = (mg: string): string => {
  const defaults: Defaults = {
    xsmall: '5',
    small: '8',
    medium: '16',
    larger: '24',
    xlarger: '32',
  }
  const value = defaults[`${mg}`] ? `${defaults[`${mg}`]}px` : ''
  return value
}

export const TextArea = styled.textarea<ITextAreaLabelStyle>`
  padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '8px')};
  border: 1px solid #00abff;
  border-color: ${(props) => props.$bdColor || ''};
  margin: 1px;
  resize: ${(props) => props.$resize};
  width: 100%;
  height: 100%;

  &:focus {
    outline: none;
    border: 2px solid #00abff;
    margin: 0;
  }
`

export const Label = styled.label`
  position: absolute;
  top: -10px;
  left: 12px;
  background-color: '';

  span {
    font-size: 14px;
    color: red;
    margin-right: 3px;
  }
`

export const Conteiner = styled.div<ITextAreaLabelStyle>`
  font-size: ${(props) => (props.$fs ? fontSize(props.$fs) : '')};
  font-style: ${(props) => props.$fStyle || ''};
  font-weight: ${(props) => (props.$fw ? fontWheigt(props.$fw) : '')};
  font-family: ${(props) => props.$fFamily || ''};
  text-align: ${(props) => props.$tAlign || ''};
  text-transform: ${(props) => props.$tTransform || ''};
  color: ${(props) => props.$color || ''};
  background-color: ${(props) => props.$bgColor || ''};
  padding: ${(props) => (props.$pd ? paddingFn(props.$pd) : '')};
  padding-top: ${(props) => (props.$pdTop ? props.$pdTop + 'px' : '')};
  padding-left: ${(props) => (props.$pdLeft ? props.$pdLeft + 'px' : '')};
  padding-right: ${(props) => (props.$pdRigth ? props.$pdRigth + 'px' : '')};
  padding-bottom: ${(props) => (props.$pdBottom ? props.$pdBottom + 'px' : '')};
  margin: ${(props) => (props.$mg ? marginFn(props.$mg) : '')};
  margin-top: ${(props) => (props.$mgTop ? props.$mgTop + 'px' : '')};
  margin-left: ${(props) => (props.$mgLeft ? props.$mgLeft + 'px' : '')};
  margin-right: ${(props) => (props.$mgRigth ? props.$mgRigth + 'px' : '')};
  margin-bottom: ${(props) => (props.$mgBottom ? props.$mgBottom + 'px' : '')};
  max-width: ${(props) => (props.$width ? props.$width + 'px' : '250px')};
  height: ${(props) => (props.$height ? props.$height + 'px' : '120px')};
  position: relative;

  &.primary {
    ${TextArea} {
      border-radius: 8px;
    }
  }

  &.secondary {
    ${TextArea} {
      border-radius: 8px 0;
    }
  }

  &.error {
    color: ${(props) => props.$color || '#ff0000'};
    ${TextArea} {
      border: 1px solid #ff0000;
      &:focus {
        border: 2px solid #ff0000;
      }
    }
  }
`
