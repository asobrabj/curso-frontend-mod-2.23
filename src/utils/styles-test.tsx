import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
