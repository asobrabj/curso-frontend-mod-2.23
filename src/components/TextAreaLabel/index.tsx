import * as S from './styles'

export interface ITextAreaLabel {
  fs?: 'xsmall' | 'small' | 'medium' | 'larger' | string
  fStyle?: 'normal' | 'italic'
  fw?: 'medium' | 'larger'
  fFamily?: string
  tAlign?: 'start' | 'end' | 'justify' | 'center'
  tTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase'
  color?: string
  bgColor?: string
  pd?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  pdTop?: string
  pdBottom?: string
  pdRigth?: string
  pdLeft?: string
  mg?: 'xsmall' | 'small' | 'medium' | 'larger' | 'xlarger' | 'none'
  mgTop?: string
  mgBottom?: string
  mgRigth?: string
  mgLeft?: string
  className?: string
  width?: string
  height?: string
  placeholder?: string
  resize?: 'both' | 'none' | 'vertical' | 'horizontal'
  bdColor?: string
  required: boolean
  label: string
  id: string
  onChange?: (e: React.ChangeEvent<HTMLTextAreaElement>) => void
  value?: string
  title?: string
}

export default function TextAreaLabel({
  fs,
  fStyle,
  fw,
  fFamily,
  tAlign,
  tTransform,
  color,
  bgColor,
  pd,
  pdTop,
  pdRigth,
  pdLeft,
  pdBottom,
  mg,
  mgTop,
  mgRigth,
  mgLeft,
  mgBottom,
  className,
  width,
  height,
  placeholder,
  resize = 'none',
  bdColor,
  required = false,
  label,
  id,
  onChange,
  value,
  title,
}: ITextAreaLabel) {
  return (
    <S.Conteiner
      className={`primary ${className}`}
      $fs={fs}
      $fw={fw}
      $fStyle={fStyle}
      $fFamily={fFamily}
      $tAlign={tAlign}
      $tTransform={tTransform}
      $color={color}
      $bgColor={bgColor}
      $pd={pd}
      $pdTop={pdTop}
      $pdRigth={pdRigth}
      $pdLeft={pdLeft}
      $pdBottom={pdBottom}
      $mg={mg}
      $mgTop={mgTop}
      $mgRigth={mgRigth}
      $mgLeft={mgLeft}
      $mgBottom={mgBottom}
      $width={width}
      $height={height}
    >
      <S.Label htmlFor={id}>
        {required && <span>*</span>}
        {label}
      </S.Label>

      <S.TextArea
        id={id}
        required={required}
        $resize={resize}
        $bdColor={bdColor}
        placeholder={placeholder}
        onChange={onChange}
        name={id}
        title={title}
        value={value}
      />
    </S.Conteiner>
  )
}
