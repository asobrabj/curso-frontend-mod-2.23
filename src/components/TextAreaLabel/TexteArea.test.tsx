import { fireEvent } from '@testing-library/react'
import TextAreaLabel, { ITextAreaLabel } from '.'
import { renderTheme } from '../../utils/styles-test'

const mockOnChange = jest.fn()

const defaultProps: ITextAreaLabel = {
  required: false,
  label: 'Test Label',
  id: 'test-id',
  onChange: mockOnChange,
}

describe('TextAreaLabel Component', () => {
  test('Renderiza corretamente', () => {
    const { getByLabelText } = renderTheme(<TextAreaLabel {...defaultProps} />)

    expect(getByLabelText('Test Label')).toBeInTheDocument()
  })

  test('Chama a função onChange corretamente', () => {
    const { getByLabelText } = renderTheme(<TextAreaLabel {...defaultProps} />)
    const textArea = getByLabelText('Test Label')

    fireEvent.change(textArea, { target: { value: 'Texto de teste' } })

    expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object))
  })

  test('Renderiza o indicador de campo obrigatório', () => {
    const { getByText } = renderTheme(
      <TextAreaLabel {...defaultProps} required />
    )

    expect(getByText('*')).toBeInTheDocument()
  })
  it('snapshot', () => {
    const { container } = renderTheme(<TextAreaLabel />)

    expect(container).toMatchSnapshot()
  })
})
