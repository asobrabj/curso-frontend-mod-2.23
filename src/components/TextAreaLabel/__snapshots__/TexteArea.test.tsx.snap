// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaLabel Component snapshot 1`] = `
.c3 {
  padding: 8px;
  border: 1px solid #00abff;
  margin: 1px;
  resize: none;
  width: 100%;
  height: 100%;
}

.c3:focus {
  outline: none;
  border: 2px solid #00abff;
  margin: 0;
}

.c1 {
  position: absolute;
  top: -10px;
  left: 12px;
  background-color: '';
}

.c1 span {
  font-size: 14px;
  color: red;
  margin-right: 3px;
}

.c0 {
  max-width: 250px;
  height: 120px;
  position: relative;
}

.c0.primary .c2 {
  border-radius: 8px;
}

.c0.secondary .c2 {
  border-radius: 8px 0;
}

.c0.error {
  color: #ff0000;
}

.c0.error .c2 {
  border: 1px solid #ff0000;
}

.c0.error .c2:focus {
  border: 2px solid #ff0000;
}

<div>
   
  <div
    class="c0 primary undefined"
  >
    <label
      class="c1"
    />
    <textarea
      class="c2 c3"
    />
  </div>
   
</div>
`;
